### DISCUSSION TITLE
[Typescript] Dynamic import not working with custom hook

### DISCUSSION THREAD

**Discussion by stefanbajic94** [vercel/next.js#16987] [Help] [answered]
I'm trying to dynamically import a custom hook which loads state from local storage.

`hooks/userStateWithLocalStorage.tsx`
```
import React from 'react'

const useStateWithLocalStorage = (localStorageKey: string, defaultValue: any = ''): [string, React.Dispatch<React.SetStateAction<string>>] => {
  const [value, setValue] = React.useState(
    localStorage.getItem(localStorageKey) || String(defaultValue)
  )

  React.useEffect(() => {
    localStorage.setItem(localStorageKey, value)
  }, [value])

  return [value, setValue]
}

export default useStateWithLocalStorage

```

`components/Calculator/Calculator.tsx`
```
...
const useStateWithLocalStorage = dynamic(() => import('../../hooks/useStateWithLocalStorage'), { ssr: false })
...
```

I'm getting a TypeError: useStateWithLocalStorage is not a function or its return value is not iterable

**Comment by jamesmosier** [comment, marked as answer]
Hi @x4th. `next/dynamic` only works with React components. You could try to just use `import()` instead. There is an upcoming release which update the docs, but for now here is a PR with the updates that may be helpful: https://github.com/vercel/next.js/pull/16803/files

The general idea would be:

```js
const useStateWithLocalStorage = await import('../../hooks/useStateWithLocalStorage').default
```

**Reply by stefanbajic94** [reply]
Yeah, the current page on dynamic imports states "With it you can import JavaScript modules (inc. React Components) dynamically and work with them." which makes it seem that it's not only limited to React components. If I do what you've suggested, is SSR off for that function i.e. it'll only get executed client-side?

**Reply by jamesmosier** [reply]
According to the new docs that will be incoming : `They also work with SSR`. But you'd have to test your specific use case to be certain how it responds.

**Reply by AdrianMrn** [reply]
@x4th Could you post an example of how you managed to fix your issue in the end? I can import my custom hook, but when using it, React throws an error `Rendered more hooks than during the previous render.`. This is what I have so far (I realise it looks a bit silly):

```js
async function loadCustomHook() {
    const customHook = await import('../customHook');

    return customHook;
}

export default function Home() {
    const [useCustomHook, setUseCustomHook] = useState(() => () => {});

    useEffect(() => {
        loadCustomHook().then(customHook => setUseCustomHook(customHook));
    }, []);

    const foo = useCustomHook({});

```

I understand why I'm getting this error (you can't conditionally use hooks, React components expect the same amount of hooks in any given render etc), but I can't figure out how to get around it in this case.

Edit: I might have found a solution here: https://stackoverflow.com/a/63438354/6374824

**Comment by nandorojo** [comment]
@jamesmosier Let's say that I want to import `firebase` into my app without bloating it, so I want to lazy load it.

I was thinking of importing `firebase` in a custom auth provider, using `next/dynamic` to import the provider into my `_app.tsx`, and then accessing the context throughout the app. The question is, do I have to `await import('firebase/app').default` any time I use it throughout the my project, or just the first time when I do so at the root? I figure only at the root since the code has been loaded in already at that point, but I'm not certain.

Here's what I had planned:

```tsx
// provider.ts
import * as firebase from 'firebase/app'
import 'firebase/auth'

export default function AuthProvider() {
  return ...
}

export const useAuthContext = () => ...
```

```tsx
// pages/_app.tsx
import dynamic from 'next/dynamic'
const AuthProvider = dynamic(() => import('../provider'), { ssr: false })

export default function App({ Component, pageProps }) {
  return (
    <AuthProvider>
      <Component {...pageProps} />
    </AuthProvider>
  )
}
```

The question is, if I want to use `firebase` throughout my app, do I have to lazy-load every file that relies on it?

```tsx
import firebase from 'firebase/app'

const getUsers = async () => {
  const token = await firebase.auth().currentUser?.getIdToken()
  ...
}

// should it look like this?
const getUsers = async () => {
  const firebase = await import('firebase/app').default
  const token = await firebase.auth().currentUser?.getIdToken()
  ...
}
```

Thanks!

**Reply by jamesmosier** [reply]
There is no benefit that I see to using `next/dynamic` for a Context Provider inside `_app`. What benefit do you see in using it? 

Usually `next/dynamic` is best used in pages/components that want to important a singular component.

**Reply by nandorojo** [reply]
Got it, makes sense. So if I want to have my user’s  authentication status in the top right of the header, which shows up on every page, there’s no benefit from any sort of lazy loading, right?

**Reply by jamesmosier** [reply]
Correct. If down the road you feel it is somehow slowing your site down then you can look at options for optimizing it. For now I think your site should run just fine with it not dynamically imported 😄

**Reply by nandorojo** [reply]
Thanks! For future reference, what other optimizations could exist for this scenario besides lazy loading?

**Reply by nandorojo** [reply]
Hey just following up here.

I tried using `next/dynamic` in my provider for my firebase auth context. I built once doing that, and once importing it normally.

I then ran `next build` in both cases to compare. When I use `next/dynamic` (with `ssr: false`), my `/_app` "First Load JS" is **217kb**. When I import normally, it goes up to **487kb**. 

Is this relevant? I might be unfamiliar with what the "First Load JS" means, since it will still require the code when the client loads anyway. But it feels like a pretty substantial difference. I appreciate your help here, thanks for taking the time.

**Reply by kartikxisk** [reply]
"First Load JS" it will share this js file across your pages that you have created. it will load on each pages. 
and next/dynamic, what it does
It create a chunk for dynamic loaded file. it will not load at the the main page file it load separately
which eventually speed up your app
